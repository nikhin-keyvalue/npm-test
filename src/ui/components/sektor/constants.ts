const SECTOR_COLOR = '#FFFFFF';
const CIRCLE_COLOR = '#4F4F4F';

export { SECTOR_COLOR, CIRCLE_COLOR };
