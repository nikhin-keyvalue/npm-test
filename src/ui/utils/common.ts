import { ContractInterface, ethers } from 'ethers';
import { Cluster } from '@solana/web3.js';

import { AddEthereumChainParameter } from 'types/wallet';
import { USDCAbi, USDTAbi } from 'constants/abi';
import {
  MUMBAI_CHAIN_PARAMS,
  TokenTypes,
  EthChains,
  POLYGON_CHAIN_PARAMS,
  SolChains,
  SolanaCluster,
  ANDROID_WALLETCONNECT_DEEPLINK,
  CurrencyTypes,
  WalletConnectWalletNames
} from 'constants/wallet';
import { OS, REDIRECT_URLS, Wallets, WALLET_ICONS } from 'constants/common';
import { AvailableNetworks, ConnectionInformation, WalletData, WalletDetails } from 'types';
import { MOBILE_DEVICE_REGEX } from 'constants/regex';
import { getWalletMeta } from 'ui/wallets/WalletConnect';
import { ASSETS_PATH } from 'config';

const parseChain = (chain: EthChains | number) =>
  typeof chain === 'number' ? (`0x${chain.toString(16)}` as EthChains) : chain;

const parseAddress = (address: string) =>
  `${address.slice(0, 2)}${address.slice(2, 6).toUpperCase()}...${address.slice(-4).toUpperCase()}`;

const parseDecimals = (decimals: ethers.BigNumber | number) =>
  typeof decimals === 'number' ? decimals : decimals.toNumber();

const getChainParams = (chain: EthChains) => {
  if (chain === EthChains.MUMBAI) return MUMBAI_CHAIN_PARAMS;
  if (chain === EthChains.POLYGON) return POLYGON_CHAIN_PARAMS;

  return {} as AddEthereumChainParameter;
};

const getContractAbi = (token: TokenTypes) => {
  const contractAbi: ContractInterface = token === TokenTypes.USDT ? USDTAbi : USDCAbi;

  return contractAbi;
};

const getClusterFromChain = (chain: SolChains): Cluster => {
  if (chain === SolChains.DEVNET) return SolanaCluster.DEVNET;

  return SolanaCluster.MAINNET;
};

const getWalletName = (walletName: Wallets) => {
  switch (walletName) {
    case Wallets.METAMASK:
      return 'metamask';
    case Wallets.PHANTOM:
      return 'phantom';
    default:
      return '';
  }
};

const handleWalletInstallRedirect = (wallet: Wallets) => {
  const url = REDIRECT_URLS[wallet];

  window.open(url);
};

const isMobileDevice = () => MOBILE_DEVICE_REGEX.test(navigator.userAgent);

const triggerAndroidWalletDeepLink = () => {
  window.open(ANDROID_WALLETCONNECT_DEEPLINK);
};

const getLocalStorageWalletConnect = () => window.localStorage.getItem('walletconnect');

const isWalletConnect = (wallet: WalletData) => wallet.name === Wallets.WALLETCONNECT;

const isAndroid = (deviceOS: string) => deviceOS === OS.ANDROID;

const isIOS = (deviceOS: string) => deviceOS === OS.IOS;

const isTrustWallet = (walletName: string) =>
  walletName === WalletConnectWalletNames.TRUST_ANDROID || walletName === WalletConnectWalletNames.TRUST_IOS;

//TODO: image need to be fetched from BE
const getSelectedWalletIcon = (selectedWallet: WalletData) => {
  const walletMeta = getWalletMeta();

  // TODO: fallback for all popular wallets
  if (isWalletConnect(selectedWallet))
    return (
      (walletMeta?.icons && walletMeta?.icons[0]) ||
      `${ASSETS_PATH}/public/images/${
        isTrustWallet(walletMeta?.name || '') ? 'trust-wallet' : 'wallet-connect'
      }-icon.svg`
    );

  return `${ASSETS_PATH}/public/images/${getWalletName(selectedWallet.name)}-icon.svg`;
};

const getSelectedWalletName = (selectedWallet: WalletData) =>
  isWalletConnect(selectedWallet) ? getWalletMeta()?.name : selectedWallet.name;

const getFallbackWalletIcon = (wallet: Wallets) => `${ASSETS_PATH}/public/images/${WALLET_ICONS[wallet]}.svg`;

// TODO: compare with function in bitcoin branch
const copyToClipBoard = (copyText: string) => {
  navigator.clipboard.writeText(copyText);
};

const getBTCConnectionInfo = (availableNetworks: AvailableNetworks[], paymentOption: WalletDetails) => {
  const BTCNetwork =
    availableNetworks.find((item) => item.network_code === CurrencyTypes.BTC)?.network_code || CurrencyTypes.BTC;
  const connectionInfo: ConnectionInformation = {
    network_code: BTCNetwork,
    payment_option_id: paymentOption[Wallets.BTC].id
  };

  return connectionInfo;
};

export {
  parseAddress,
  parseChain,
  parseDecimals,
  getChainParams,
  getContractAbi,
  getClusterFromChain,
  getWalletName,
  getFallbackWalletIcon,
  handleWalletInstallRedirect,
  isMobileDevice,
  triggerAndroidWalletDeepLink,
  getLocalStorageWalletConnect,
  isWalletConnect,
  isAndroid,
  isIOS,
  getSelectedWalletIcon,
  getSelectedWalletName,
  copyToClipBoard,
  getBTCConnectionInfo
};
