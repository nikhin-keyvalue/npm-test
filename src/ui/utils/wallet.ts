import { ethers } from 'ethers';

import { Wallets } from 'constants/common';
import { CoinData } from 'types';
import { CurrencyTypes, EthChains, SolChains, TokenTypes } from 'constants/wallet';
import { Balance, ChainTypes, CoinTypes } from 'types/wallet';
import { getEthBalance, getTokenBalance } from '../wallets/common';
import { getSolBalance, getSolTokenBalance } from '../wallets/Phantom';

const isSolChain = (chain: ChainTypes): chain is SolChains => chain === SolChains.DEVNET || chain === SolChains.MAINNET;

const isEthChain = (chain: ChainTypes): chain is EthChains =>
  chain === EthChains.MAINNET ||
  chain === EthChains.RINKEBY ||
  chain === EthChains.POLYGON ||
  chain === EthChains.MUMBAI;

const isEthereumBasedCoin = (coin: CoinTypes): coin is CurrencyTypes.ETHEREUM | CurrencyTypes.MATIC =>
  coin === CurrencyTypes.ETHEREUM || coin === CurrencyTypes.MATIC;

const isSolBasedCoin = (coin: CoinTypes): coin is CurrencyTypes.SOL => coin === CurrencyTypes.SOL;

const isToken = (coin: CoinTypes): coin is TokenTypes => coin === TokenTypes.USDC || coin === TokenTypes.USDT;

const getCurrencyFromChain = (chain: ChainTypes) => {
  if (chain === EthChains.MAINNET || chain === EthChains.RINKEBY) return CurrencyTypes.ETHEREUM;
  else if (chain === EthChains.POLYGON || chain === EthChains.MUMBAI) return CurrencyTypes.MATIC;
  else if (chain === SolChains.DEVNET || chain === SolChains.MAINNET) return CurrencyTypes.SOL;

  return '' as CurrencyTypes;
};

const getEthBalancesArray = async (
  address: string,
  tokenList: CoinData[],
  chain: EthChains,
  selectedWallet: Wallets
) => {
  const balance = await getEthBalance(address, selectedWallet);
  const mainBalance: Balance = {
    coin: getCurrencyFromChain(chain),
    balance: Number(ethers.utils.formatEther(balance))
  };

  const tokenBalances = await Promise.all(
    tokenList.map(async (item) => {
      if (item) {
        const tokenBalance = await getTokenBalance(
          address,
          item.currency_code as TokenTypes,
          chain,
          item.contract_address,
          selectedWallet
        );

        return {
          coin: item.currency_code,
          balance: tokenBalance
        } as Balance;
      }

      return {
        coin: null,
        balance: 0
      };
    })
  );

  return [mainBalance, ...tokenBalances];
};

const getSolBalancesArray = async (address: string, tokenList: CoinData[], chain: SolChains) => {
  const balance = await getSolBalance(address, chain);
  const mainBalance: Balance = {
    coin: getCurrencyFromChain(chain),
    balance
  };

  const tokenBalances = await Promise.all(
    tokenList.map(async (item) => {
      if (item) {
        const tokenBalance = await getSolTokenBalance(address, item.contract_address, chain);

        return {
          coin: item.currency_code,
          balance: tokenBalance
        } as Balance;
      }

      return {
        coin: null,
        balance: 0
      };
    })
  );

  return [mainBalance, ...tokenBalances];
};

const isValidBalance = (balance: Balance[]) => {
  if (balance.length) {
    // TODO: make function
    const coins: CoinTypes[] = [...Object.values(CurrencyTypes), ...Object.values(TokenTypes)];
    let valid = true;

    balance.forEach((item) => {
      if (!(item.coin && coins.includes(item.coin))) valid = false;

      return;
    });

    return valid;
  } else {
    return false;
  }
};

export {
  isSolChain,
  isEthChain,
  isEthereumBasedCoin,
  isSolBasedCoin,
  isToken,
  getCurrencyFromChain,
  getEthBalancesArray,
  getSolBalancesArray,
  isValidBalance
};
