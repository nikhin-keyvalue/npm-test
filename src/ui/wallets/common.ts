import WalletConnectProvider from '@walletconnect/web3-provider';
import { ethers } from 'ethers';

import { Wallets } from 'constants/common';
import { ContractFunctions, EthChains, TokenTypes } from 'constants/wallet';
import { getContractAbi, parseDecimals } from '../utils/common';
import { getMetaMaskProvider } from './Metamask';
import { getWalletConnectProvider } from './WalletConnect';

const getContractData = (contract: ethers.Contract, functionName: ContractFunctions, params: any[]) => {
  const iface = contract.interface;
  const data = iface.encodeFunctionData(functionName, params);

  return data;
};

const getWeb3ProviderAndSigner = (provider: ethers.providers.ExternalProvider) => {
  const web3provider = new ethers.providers.Web3Provider(provider);
  const signer = web3provider.getSigner();

  return {
    web3provider,
    signer
  };
};

const getTokenDecimals = async (contract: ethers.Contract) => {
  const decimalsResp: ethers.BigNumber = await contract.decimals();

  return parseDecimals(decimalsResp);
};

const getEthProvider = (wallet: Wallets) => {
  switch (wallet) {
    case Wallets.METAMASK:
      return getMetaMaskProvider();
    case Wallets.WALLETCONNECT:
      return getWalletConnectProvider();
    default:
      return null;
  }
};

const getEthBalance = async (address: string, wallet: Wallets) => {
  try {
    const provider: ethers.providers.ExternalProvider | WalletConnectProvider = getEthProvider(wallet);

    const { web3provider } = getWeb3ProviderAndSigner(provider);
    const balance = await web3provider.getBalance(address);

    return balance;
  } catch (err: any) {
    return 0;
  }
};

const getTokenBalance = async (
  address: string,
  token: TokenTypes,
  chain: EthChains,
  contractAddress: string,
  wallet: Wallets
) => {
  try {
    const provider: ethers.providers.ExternalProvider | WalletConnectProvider = getEthProvider(wallet);
    const { signer } = getWeb3ProviderAndSigner(provider);

    const contractAbi = getContractAbi(token);
    const contract = new ethers.Contract(contractAddress, contractAbi, signer);

    const balanceResp: ethers.BigNumber = await contract.balanceOf(address);
    const decimals = await getTokenDecimals(contract);
    // TODO: toNumber breaks when balance to large
    const balance = balanceResp.toNumber();

    return balance / Math.pow(10, decimals);
  } catch (err: any) {
    console.log(err);

    return 0;
  }
};

export { getContractData, getWeb3ProviderAndSigner, getTokenDecimals, getEthBalance, getTokenBalance };
