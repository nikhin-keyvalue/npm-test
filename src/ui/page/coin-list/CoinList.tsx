import Preact, { h } from 'preact';
import { useState } from 'preact/hooks';

import PaymentDetails from 'ui/components/payment-details/PaymentDetails';
import Button from 'ui/components/button/Button';
import PaymentDetailsBottomSheet from 'ui/components/bottom-sheet/payment-details/paymentdetails';
import BottomSheet from 'ui/components/bottom-sheet/BottomSheet';
import { getCurrentPhantomAccount, sendSol, transferSolToken } from 'ui/wallets/Phantom';
import { getWalletConnectAccount, transferTokenWalletConnect, WalletConnectSendEth } from 'ui/wallets/WalletConnect';
import { isIOS, isMobileDevice } from 'ui/utils/common';
import { getMetamaskAccount, sendEth, transferToken } from 'ui/wallets/Metamask';
import { BottomSheetTypes, ButtonType, ModalTypes, Wallets, ZampfiPages } from 'constants/common';
import { CoinTypes, TransferOrSendResponse, SolTransferOrSendResponse } from 'types/wallet';
import { CoinData, CoinListProps, PostPaymentInfo } from 'types';
import { isEthereumBasedCoin, isSolBasedCoin, isSolChain, isToken } from 'ui/utils/wallet';
import { getDeviceOS } from 'zamp/payment/utils';
import { ASSETS_PATH } from 'config';

import './style.scss';

const CoinList: Preact.FunctionComponent<CoinListProps> = ({
  handlePageChange,
  setSelectedCoinObject,
  setSelectedAmount,
  onTransactionCompleted,
  selectedChain,
  address,
  convertedAmounts,
  amount,
  sourceCurrency,
  isPaymentStarted,
  setIsPaymentStarted,
  onCoinSelect,
  balances,
  setModalOpen,
  selectedWallet,
  isBottomSheetOpen,
  setBottomSheetOpen,
  setCurrentPage
}) => {
  const [showPaymentDetails, setShowPaymentDetails] = useState<string>();

  const handleInvalidAccount = () => {
    setModalOpen(ModalTypes.INVALID_ACCOUNT);
    setIsPaymentStarted(false);
  };

  const onPayButtonClick = async (item: CoinData) => {
    setBottomSheetOpen(null);
    setSelectedCoinObject(item);
    setSelectedAmount(item.amount);
    if (selectedWallet.name === Wallets.BINANCE) {
      setCurrentPage(ZampfiPages.BINANCE_SCREEN);
    } else {
      setIsPaymentStarted(true);
      const res = await onCoinSelect(item.currency_code);

      if (res?.data) {
        await sendTransaction(res.data, item);
      } else {
        setIsPaymentStarted(false);
        setCurrentPage(ZampfiPages.ERROR_SCREEN);
      }
    }
  };

  const sendTransaction = async (paymentDetails: PostPaymentInfo, coinInfo: CoinData) => {
    const { dest_account: toAddress, id: paymentId, amount, lowest_denom_amount, currency_code: coin } = paymentDetails;

    let response: TransferOrSendResponse = {
      success: false,
      transactionResponse: null
    };

    let solResponse: SolTransferOrSendResponse = {
      success: false,
      transactionResponse: null
    };

    switch (selectedWallet.name) {
      case Wallets.METAMASK:
        setModalOpen(ModalTypes.WALLET_CONFIRM);
        if ((await getMetamaskAccount()) !== address) {
          handleInvalidAccount();

          return;
        }
        if (isEthereumBasedCoin(coin)) response = await sendEth(address as string, toAddress, amount);
        else if (isToken(coin))
          response = await transferToken(toAddress, lowest_denom_amount, coin, coinInfo.contract_address);

        if (response?.transactionResponse) {
          await onTransactionCompleted(response.transactionResponse?.hash, paymentId);

          handlePageChange(ZampfiPages.SUCCESS_SCREEN);
        } else if (!response.success) {
          handlePageChange(ZampfiPages.ERROR_SCREEN);
        }
        break;

      case Wallets.WALLETCONNECT:
        if (!isMobileDevice() || isIOS(getDeviceOS())) setModalOpen(ModalTypes.WALLET_CONFIRM);
        if (getWalletConnectAccount() !== address) {
          handleInvalidAccount();

          return;
        }
        if (isEthereumBasedCoin(coin)) response = await WalletConnectSendEth(toAddress, amount);
        else if (isToken(coin))
          response = await transferTokenWalletConnect(toAddress, lowest_denom_amount, coin, coinInfo.contract_address);

        if (response?.transactionResponse) {
          await onTransactionCompleted(response.transactionResponse?.hash, paymentId);

          handlePageChange(ZampfiPages.SUCCESS_SCREEN);
        } else if (!response.success) {
          // TODO: walletConnect rejection must be handled as a notifier
        }
        break;
      case Wallets.PHANTOM:
        // TODO: might change when implementing mobile flow
        setModalOpen(ModalTypes.WALLET_CONFIRM);
        if (getCurrentPhantomAccount() !== address) {
          handleInvalidAccount();

          return;
        }
        if (isSolChain(selectedChain))
          if (isSolBasedCoin(coin))
            solResponse = await sendSol(address as string, toAddress, lowest_denom_amount, selectedChain);
          else if (isToken(coin))
            solResponse = await transferSolToken(
              toAddress,
              lowest_denom_amount,
              selectedChain,
              coinInfo.contract_address
            );

        if (solResponse.transactionResponse) {
          await onTransactionCompleted(solResponse.transactionResponse, paymentId);

          handlePageChange(ZampfiPages.SUCCESS_SCREEN);
        } else if (!solResponse.success) {
          handlePageChange(ZampfiPages.ERROR_SCREEN);
        }
    }

    setIsPaymentStarted(false);
    setModalOpen(null);
  };

  const handlePaymentDetailsClick = (coinId: string) => {
    if (isMobileDevice()) setBottomSheetOpen(BottomSheetTypes.PAYMENT_DETAILS);
    if (coinId === showPaymentDetails) setShowPaymentDetails('');
    else setShowPaymentDetails(coinId);
  };

  const hasBalance = (coin: CoinTypes, amount: string) =>
    Boolean((balances.find((balanceItem) => balanceItem.coin === coin)?.balance || 0) >= Number(amount));

  return (
    <div className='zamp-coin-container'>
      {convertedAmounts.map((item) => (
        <div className='zamp-coin-wrapper' key={item.currency_code}>
          <div className='zamp-coin-info'>
            <div className='zamp-coin-item'>
              <div className='zamp-coin-details'>
                <div>
                  <img src={item?.logo_url} alt='Coin' height='20' />
                </div>
                <div>
                  <div className='zamp-coin-name'>{item.display_name}</div>
                  <div
                    className='zamp-coin-value'
                    role='presentation'
                    onClick={() => handlePaymentDetailsClick(item.currency_code)}
                  >
                    {`${Number(item.amount).toFixed(10)} ${item.currency_code}`}
                    <img
                      src={`${ASSETS_PATH}/public/images/right-arrow.svg`}
                      alt='Arrow icon'
                      className='zamp-coin-arrow-icon'
                      style={{
                        transform:
                          showPaymentDetails === item.currency_code && !isMobileDevice()
                            ? 'rotate(270deg)'
                            : 'rotate(90deg)'
                      }}
                    />
                  </div>
                </div>
              </div>
              <div
                className={`zamp-coin-list-button-container ${
                  selectedWallet.name !== Wallets.BINANCE &&
                  !hasBalance(item.currency_code, item.amount) &&
                  'zamp-coin-list-insufficient-fund-indicator-button'
                }`}
                role='presentation'
              >
                <Button
                  type={ButtonType.SECONDARY}
                  buttonLabel='Pay now'
                  width='73px'
                  height='32px'
                  handleButtonClick={() => onPayButtonClick(item)}
                  //TODO: make function
                  isDisabled={
                    selectedWallet.name !== Wallets.BINANCE &&
                    (isPaymentStarted || !hasBalance(item.currency_code, item.amount))
                  }
                />
                <div className='zamp-insufficient-fund-indicator-container'>Low Balance</div>
              </div>
            </div>
            <PaymentDetails
              isVisible={showPaymentDetails === item?.currency_code && !isMobileDevice()}
              totalPrice={amount}
              exchangeRate={item.exchange_rate}
              subTotal={item.amount}
              sourceCurrency={sourceCurrency}
              currencyCode={item.currency_code}
            />
            <BottomSheet
              isOpen={Boolean(isBottomSheetOpen) && showPaymentDetails === item?.currency_code}
              setBottomSheetOpen={setBottomSheetOpen}
            >
              <PaymentDetailsBottomSheet
                totalPrice={amount}
                subTotal={item.amount}
                sourceCurrency={sourceCurrency}
                currencyCode={item.currency_code}
                coinName={item.display_name}
                exchangeRate={item.exchange_rate}
                coinLogourl={item.logo_url}
                handlePayButtonClick={() => onPayButtonClick(item)}
                isPayButtonDisabled={
                  selectedWallet.name !== Wallets.BINANCE &&
                  (isPaymentStarted || !hasBalance(item.currency_code, item.amount))
                }
              />
            </BottomSheet>
          </div>
        </div>
      ))}
    </div>
  );
};

export default CoinList;
