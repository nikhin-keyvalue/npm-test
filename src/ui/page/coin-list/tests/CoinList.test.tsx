// TODO: remove
/* eslint-disable @typescript-eslint/no-unused-vars */

import { h } from 'preact';
import { configure, mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { Adapter } from 'enzyme-adapter-preact-pure';

import CoinList from 'ui/page/coin-list/CoinList';
import * as metamask from 'ui/wallets/Metamask';
import { CurrencyTypes, EthChains } from 'constants/wallet';
import { CoinData, WalletData, WalletDetails } from 'types';
import { TransferOrSendResponse } from 'types/wallet';
import { Wallets } from 'constants/common';

configure({ adapter: new Adapter() });

describe('CoinList', () => {
  const amountWithExchangeRate: CoinData[] = [
    {
      amount: '0',
      exchange_rate: '0',
      currency_code: CurrencyTypes.ETHEREUM,
      display_name: 'Ethereum',
      type: 'crypto',
      logo_url: '',
      contract_address: '',
      lowest_denom_amount: ''
    }
  ];

  const walletDetails: WalletData = {
    id: '',
    logo_url: '',
    name: Wallets.METAMASK,
    type: ''
  };

  const mockFn = jest.fn();

  jest.useFakeTimers();
  it('render components', () => {
    const wrapper = mount(
      <CoinList
        handlePageChange={mockFn}
        setSelectedCoinObject={mockFn}
        setSelectedAmount={mockFn}
        onCoinSelect={mockFn}
        onTransactionCompleted={mockFn}
        currentChain={EthChains.RINKEBY}
        selectedChain={EthChains.RINKEBY}
        amount={'0'}
        sourceCurrency={'USD'}
        address={'0x1234'}
        balances={[]}
        convertedAmounts={amountWithExchangeRate}
        setModalOpen={mockFn}
        selectedWallet={walletDetails}
        isPaymentStarted={false}
        setIsPaymentStarted={mockFn}
        isBottomSheetOpen={null}
        setBottomSheetOpen={() => null}
        setCurrentPage={mockFn}
      />
    );
    const tree = shallowToJson(wrapper);

    expect(tree).toMatchSnapshot();
  });

  it('handle pay Coins click', async () => {
    const onHandleCoinsPageMock = jest.fn();
    const setSelectedCoinMock = jest.fn();
    const wrapper = shallow(
      <CoinList
        handlePageChange={onHandleCoinsPageMock}
        setSelectedAmount={mockFn}
        setSelectedCoinObject={mockFn}
        onCoinSelect={mockFn}
        onTransactionCompleted={mockFn}
        currentChain={EthChains.RINKEBY}
        selectedChain={EthChains.RINKEBY}
        amount={'0'}
        sourceCurrency={'USD'}
        address={'0x1234'}
        convertedAmounts={amountWithExchangeRate}
        balances={[]}
        setModalOpen={mockFn}
        selectedWallet={walletDetails}
        isPaymentStarted={false}
        setIsPaymentStarted={mockFn}
        isBottomSheetOpen={null}
        setBottomSheetOpen={() => null}
        setCurrentPage={mockFn}
      />
    );
    const sendEthMock = jest.spyOn(metamask, 'sendEth');

    sendEthMock.mockImplementation(async () => {
      return {
        success: true,
        transactionResponse: {
          hash: '0x1234',
          confirmations: 1,
          from: '0x1234'
        }
      } as TransferOrSendResponse;
    });

    const transferTokenMock = jest.spyOn(metamask, 'transferToken');

    transferTokenMock.mockImplementation(async () => {
      return {
        success: true,
        transactionResponse: {
          hash: '0x1234',
          confirmations: 1,
          from: '0x1234'
        }
      } as TransferOrSendResponse;
    });

    const switchChainMock = jest.spyOn(metamask, 'switchChain');
    const addChainMock = jest.spyOn(metamask, 'addChain');

    switchChainMock.mockImplementation(async () => true);
    addChainMock.mockImplementation(async () => true);

    // pay ETH
    // await wrapper.find(Button).at(0).props().handleButtonClick();

    // pay USDT
    // await wrapper.find(Button).at(2).props().handleButtonClick();
    jest.advanceTimersByTime(50000);

    // expect(onHandleCoinsPageMock).toHaveBeenCalled();
    // expect(setSelectedCoinMock).toHaveBeenCalled();
  });

  it('handle pay MATIC click', async () => {
    const onHandleCoinsPageMock = jest.fn();
    const setSelectedCoinMock = jest.fn();
    const handlePaymentCompletedMock = jest.fn();
    const wrapper = shallow(
      <CoinList
        handlePageChange={onHandleCoinsPageMock}
        setSelectedAmount={mockFn}
        setSelectedCoinObject={mockFn}
        onCoinSelect={mockFn}
        onTransactionCompleted={mockFn}
        currentChain={EthChains.RINKEBY}
        selectedChain={EthChains.RINKEBY}
        amount={'0'}
        sourceCurrency={'USD'}
        address={'0x1234'}
        convertedAmounts={amountWithExchangeRate}
        balances={[]}
        setModalOpen={mockFn}
        selectedWallet={walletDetails}
        isPaymentStarted={false}
        setIsPaymentStarted={mockFn}
        isBottomSheetOpen={null}
        setBottomSheetOpen={() => null}
        setCurrentPage={mockFn}
      />
    );
    const sendEthMock = jest.spyOn(metamask, 'sendEth');

    sendEthMock.mockImplementation(async () => {
      return {
        success: true,
        transactionResponse: {
          hash: '0x1234',
          confirmations: 1,
          from: '0x1234'
        }
      } as TransferOrSendResponse;
    });

    const switchChainMock = jest.spyOn(metamask, 'switchChain');
    const addChainMock = jest.spyOn(metamask, 'addChain');

    switchChainMock.mockImplementation(async () => true);
    addChainMock.mockImplementation(async () => true);

    // pay MAT
    // await wrapper.find(Button).at(1).props().handleButtonClick();
  });

  it('handle pay Coins click fail', async () => {
    const mockFn = jest.fn();
    const wrapper = shallow(
      <CoinList
        handlePageChange={mockFn}
        setSelectedCoinObject={mockFn}
        setSelectedAmount={mockFn}
        onTransactionCompleted={mockFn}
        currentChain={EthChains.RINKEBY}
        selectedChain={EthChains.RINKEBY}
        onCoinSelect={mockFn}
        amount={'0'}
        sourceCurrency={'USD'}
        address={'0x1234'}
        convertedAmounts={amountWithExchangeRate}
        balances={[]}
        setModalOpen={mockFn}
        selectedWallet={walletDetails}
        isPaymentStarted={false}
        setIsPaymentStarted={mockFn}
        isBottomSheetOpen={null}
        setBottomSheetOpen={() => null}
        setCurrentPage={mockFn}
      />
    );
    const sendEthMock = jest.spyOn(metamask, 'sendEth');

    sendEthMock.mockImplementation(async () => {
      return {
        success: true,
        transactionResponse: null
      };
    });

    // pay ETH (fail)
    // await wrapper.find(Button).at(0).props().handleButtonClick();
  });
});
