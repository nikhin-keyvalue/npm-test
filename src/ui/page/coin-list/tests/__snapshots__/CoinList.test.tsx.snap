// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinList render components 1`] = `
<CoinList
  address="0x1234"
  amount="0"
  balances={Array []}
  convertedAmounts={
    Array [
      Object {
        "amount": "0",
        "contract_address": "",
        "currency_code": "ETH",
        "display_name": "Ethereum",
        "exchange_rate": "0",
        "logo_url": "",
        "lowest_denom_amount": "",
        "type": "crypto",
      },
    ]
  }
  currentChain="0x4"
  handlePageChange={[MockFunction]}
  isBottomSheetOpen={null}
  isPaymentStarted={false}
  onCoinSelect={[MockFunction]}
  onTransactionCompleted={[MockFunction]}
  selectedChain="0x4"
  selectedWallet={
    Object {
      "id": "",
      "logo_url": "",
      "name": "Metamask",
      "type": "",
    }
  }
  setBottomSheetOpen={[Function]}
  setCurrentPage={[MockFunction]}
  setIsPaymentStarted={[MockFunction]}
  setModalOpen={[MockFunction]}
  setSelectedAmount={[MockFunction]}
  setSelectedCoinObject={[MockFunction]}
  sourceCurrency="USD"
>
  <div
    className="zamp-coin-container"
  >
    <div
      className="zamp-coin-wrapper"
      key="ETH"
    >
      <div
        className="zamp-coin-info"
      >
        <div
          className="zamp-coin-item"
        >
          <div
            className="zamp-coin-details"
          >
            <div>
              <img
                alt="Coin"
                height="20"
                src=""
              />
            </div>
            <div>
              <div
                className="zamp-coin-name"
              >
                Ethereum
              </div>
              <div
                className="zamp-coin-value"
                onClick={[Function]}
                role="presentation"
              >
                0.0000000000 ETH
                <img
                  alt="Arrow icon"
                  className="zamp-coin-arrow-icon"
                  src="undefined/public/images/right-arrow.svg"
                  style={
                    Object {
                      "transform": "rotate(90deg)",
                    }
                  }
                />
              </div>
            </div>
          </div>
          <div
            className="zamp-coin-list-button-container false"
            role="presentation"
          >
            <Button
              buttonLabel="Pay now"
              handleButtonClick={[Function]}
              height="32px"
              isDisabled={false}
              type="secondary"
              width="73px"
            >
              <button
                className="zamp-secondary-button"
                disabled={false}
                onClick={[Function]}
                style={
                  Object {
                    "height": "32px",
                    "width": "73px",
                  }
                }
              >
                Pay now
              </button>
            </Button>
            <div
              className="zamp-insufficient-fund-indicator-container"
            >
              Low Balance
            </div>
          </div>
        </div>
        <PaymentDetails
          currencyCode="ETH"
          exchangeRate="0"
          isVisible={false}
          sourceCurrency="USD"
          subTotal="0"
          totalPrice="0"
        >
          <div
            className="zamp-payment-details-wrapper zamp-payment-details-hidden"
          >
            <div
              className="zamp-payment-details-price-container"
            >
              <div
                className="total-price-header"
              >
                Total Price
              </div>
              <div
                className="total-price"
              >
                0 USD
              </div>
            </div>
            <div
              className="zamp-payment-details-exchange-rate-container"
            >
              <div
                className="exchange-rate-header"
              >
                Exchange Rate
              </div>
              <div
                className="exchange-rate"
              >
                0 USD
              </div>
            </div>
            <hr />
            <div
              className="zamp-payment-details-subtotal-container"
            >
              <div
                className="subtotal-header"
              >
                Subtotal
              </div>
              <div
                className="subtotal"
              >
                0 ETH
              </div>
            </div>
          </div>
        </PaymentDetails>
        <BottomSheet
          isOpen={false}
          setBottomSheetOpen={[Function]}
        >
          <div
            className="zamp-bottom-sheet-wrapper "
            onClick={[Function]}
            role="presentation"
          >
            <div
              className="zamp-bottom-sheet-content zamp-sheet-content-close "
              onClick={[Function]}
              role="presentation"
            >
              <PaymentDetails
                coinLogourl=""
                coinName="Ethereum"
                currencyCode="ETH"
                exchangeRate="0"
                handlePayButtonClick={[Function]}
                isPayButtonDisabled={false}
                sourceCurrency="USD"
                subTotal="0"
                totalPrice="0"
              >
                <div
                  className="zamp-payment-details-bottomsheet-wrapper"
                >
                  <div
                    className="zamp-payment-details-bottomsheet-head"
                  >
                    <div
                      className="zamp-payment-details-bottomsheet-head-total-header"
                    >
                      Total Amount
                    </div>
                    <div
                      className="zamp-payment-details-bottomsheet-head-total"
                    >
                      0
                       
                      ETH
                    </div>
                    <div
                      className="zamp-payment-details-bottomsheet-head-coin-details"
                    >
                      <img
                        alt="Coin-logo"
                        height={8}
                        src=""
                        width={8}
                      />
                      <div
                        className="zamp-payment-details-bottomsheet-head-coin-name"
                      >
                        Ethereum
                      </div>
                    </div>
                  </div>
                  <div
                    className="zamp-payment-details-price-container"
                  >
                    <div
                      className="total-price-header"
                    >
                      Total Price
                    </div>
                    <div
                      className="total-price"
                    >
                      0 USD
                    </div>
                  </div>
                  <div
                    className="zamp-payment-details-exchange-rate-container"
                  >
                    <div
                      className="exchange-rate-header"
                    >
                      Exchange Rate
                    </div>
                    <div
                      className="exchange-rate"
                    >
                      0 USD
                    </div>
                  </div>
                  <hr />
                  <div
                    className="zamp-payment-details-subtotal-container"
                  >
                    <div
                      className="subtotal-header"
                    >
                      Subtotal
                    </div>
                    <div
                      className="subtotal"
                    >
                      0 ETH
                    </div>
                  </div>
                  <div
                    className="zamp-payment-details-bottomsheet-button-container"
                  >
                    <Button
                      buttonLabel="Pay now"
                      handleButtonClick={[Function]}
                      isDisabled={false}
                      type="secondary"
                      width="100px"
                    >
                      <button
                        className="zamp-secondary-button"
                        disabled={false}
                        onClick={[Function]}
                        style={
                          Object {
                            "height": undefined,
                            "width": "100px",
                          }
                        }
                      >
                        Pay now
                      </button>
                    </Button>
                  </div>
                </div>
              </PaymentDetails>
            </div>
          </div>
        </BottomSheet>
      </div>
    </div>
  </div>
</CoinList>
`;
